lib/urlwatch/reporters.py

```diff
@@ -668,18 +668,88 @@ def submit(self):
+                      'Website: %s  ' % (urlwatch.__url__,),
+                      'watched %d URLs in %d seconds' % (len(self.job_states), self.duration.seconds))
+            footer_len = sum(len(part) for part in footer) + len(footer) - 1
+
+            if max_length is not None:
+                remaining_len -= footer_len
+
+        # Calculate approximate available length per item, shared equally.
+        if max_length is not None:
+            len_per_details = remaining_len // len(details)
+
         if summary:
             yield from ('%d. %s' % (idx + 1, line) for idx, line in enumerate(summary))
             yield ''
 
         if show_details:
-            yield from details
+            if max_length is not None and len_per_details < 100:
+                yield "Details omitted due to message length."
+            else:
+                unprocessed = len(details)
+
+                for header, body in details:
+                    # Calculate the available length for the body and render it
+                    if max_length is not None:
+                        avail_length = len_per_details - len(header) - 1
+                    else:
+                        avail_length = None
+
+                    body = MarkdownReporter._format_details_body(body, avail_length)
+
+                    yield header
+                    yield body
+                    yield ''
+
+                    # Distribute the unused length into subsequent items,
+                    # unless we're at the last item already.
+                    if max_length is not None:
+                        actual_length = len(header) + len(body)
+                        unprocessed -= 1
+
+                        unused = len_per_details - actual_length
+                        remaining_len -= len_per_details
+                        remaining_len += unused
+
+                        if unprocessed > 0:
+                            len_per_details = remaining_len // unprocessed
 
         if summary and show_footer:
-            yield from ('--- ',
-                        '%s %s, %s  ' % (urlwatch.pkgname, urlwatch.__version__, urlwatch.__copyright__),
-                        'Website: %s  ' % (urlwatch.__url__,),
-                        'watched %d URLs in %d seconds' % (len(self.job_states), self.duration.seconds))
+            yield from footer
+
+    @staticmethod
+    def _format_details_body(s, max_length):
+        wrapper_length = len("```diff\n\n```")
+
+        # Message to print when the diff is too long.
+        trim_message = "[... diff trimmed ...]"
+        trim_message_length = len(trim_message)
+
+        if max_length is None or len(s) + wrapper_length <= max_length:
+            return "```diff\n{}\n```".format(s)
+        else:
+            while len(s) + wrapper_length + trim_message_length > max_length and '\n' in s:
+                s = re.sub("\n.*$", "", s)
```

Comment:
Couldn't we also use `.rfind('\n')` on the string to find the cutoff point without having to loop and remove line-by-line? Given that we know what the maximum length is, we can probably even do it with a single call (basically "find the last `\n` that comes before `max_length - trim_message_length - wrapper_length`, right?)

Here's some untested code (might contain off-by-one errors...):

```python3

target_max_length = max_length - trim_message_length - wrapper_length
pos = s.rfind('\n', 0, target_max_length)
if pos == -1:
    # we have a single long line so just cut it short
    pos = target_max_length
s = s[:pos]

Reply:
That's better. I didn't think of it because I forgot about rfind's third argument. I'll handle it when I get some free time.

Reply:
Done.